stdlib/public/core/FixedPoint.swift.gyb

```diff
@@ -429,44 +426,62 @@ extension ${Self} {
 extension ${Self} : SignedNumber {}
 % end
 
-// Construction from other integer types
-@_transparent
-extension ${Self} {
+
 % for src_ty in all_integer_types(word_bits):
 %    srcBits = src_ty.bits
 %    srcSigned = src_ty.is_signed
 %    Src = src_ty.stdlib_name
 %    (srcSign, srcExt) = ('s', 'sext') if srcSigned else ('u', 'zext')
-%    if Self != Src:
-  public init(_ v: ${Src}) {
+%    conversionFunction = "_convert" + Src + "to" + Self
+
+@_transparent
+private func ${conversionFunction}(_ value: ${Src}) -> (value: Builtin.${BuiltinName}, error: Bool) {
 %
-    let srcNotWord = v._value
+  let srcNotWord = value._value
 %
-%       if srcBits == bits and srcSign == sign:
-    let dstNotWord = srcNotWord
+%       if srcBits == bits and srcSign == sign: # Exact same size/signedness.
+  let (result, error) = (srcNotWord, false)
 %
-%       elif srcBits == bits:
-    let tmp = Builtin.${srcSign}_to_${sign}_checked_conversion_Int${srcBits}(srcNotWord)
-    Builtin.condfail(tmp.1)
-    let dstNotWord = tmp.0
+%       elif srcBits == bits:  # Same size, switching signs.
+  let (result, error) = Builtin.${srcSign}_to_${sign}_checked_conversion_Int${srcBits}(srcNotWord)
 %
-%       elif srcBits > bits:
-    let tmp = Builtin.${srcSign}_to_${sign}_checked_trunc_Int${srcBits}_Int${bits}(srcNotWord)
-    Builtin.condfail(tmp.1)
-    let dstNotWord = tmp.0
+%       elif srcBits > bits:  # Larger input, check for truncation.
+  let (result, error) = Builtin.${srcSign}_to_${sign}_checked_trunc_Int${srcBits}_Int${bits}(srcNotWord)
 %
-%       elif srcSigned and not signed:
-    let tmp = Builtin.s_to_u_checked_conversion_Int${srcBits}(srcNotWord)
-    Builtin.condfail(tmp.1)
-    let dstNotWord = Builtin.${srcExt}_Int${srcBits}_Int${bits}(tmp.0)
+%       elif srcSigned and not signed: # Smaller size input, also switching signs.
+  let (tmp, signError) = Builtin.s_to_u_checked_conversion_Int${srcBits}(srcNotWord)
+  let (result, error) = (Builtin.${srcExt}_Int${srcBits}_Int${bits}(tmp), signError)
 %
-%       else:
-    let dstNotWord = Builtin.${srcExt}_Int${srcBits}_Int${bits}(srcNotWord)
+%       else: # Smaller size input, not switching signs.
+  let (result, error) = (Builtin.${srcExt}_Int${srcBits}_Int${bits}(srcNotWord), false)
 %       end
-%
-    self._value = dstNotWord
+%    
+  return (result, Bool(error))
+}
+
+/// Construction from ${Src}
+@_transparent
+extension ${Self} {
+
+  @_transparent
+  public init(_ value: ${Src}) {
+    let result = ${conversionFunction}(value)
+    _precondition(result.error == false, "${Src} value cannot be converted to ${Self} without truncation.")
```

Comment:
Using `_precondition` here will lead to a massive performance regression I believe.  Please continue using `Builtin.condfail`.